### ISSUE TITLE
Touch select is not working properly on Firefox 60 or above

### ISSUE THREAD

**Issue by Bllacky** [select2/select2#5457]
The issue is rather simple and is also present on select2 demo page:
https://select2.org/data-sources/ajax

If I access this page on my mobile phone using the latest version of Mozilla Firefox, the touch event will not trigger correctly. It will register my touch event about 20px above where I actually touch the screen.

So if I touch on the display here on QuantConnect: 
![image](https://user-images.githubusercontent.com/6762165/52858304-6be7f400-3132-11e9-9262-95514bf84955.png)
It will select fable-compiler instead.

I've tested this with various browsers and devices. The issue is present only on Firefox mobile and on any device.

This issue might be related to this pull request: https://github.com/select2/select2/pull/4492

**Comment by Bllacky** [comment]
This issue is still present and reproducible.

**Comment by Bllacky** [comment]
With the latest version of select2 and firefox, the issue is still present.

**Comment by kevin-brown** [comment]
I cannot reproduce this issue on the examples page on Firefox 68 (Android) with Select2 4.0.7. I also cannot reproduce it on Chrome (Android) 75.

While it's highly unlikely that this was an issue fixed in Firefox 68 (but present in 67), can you just confirm once again for me that this issue is still present when you test it?

**Comment by Bllacky** [comment]
I just tested this on the latest version of Firefox with my website and your website. The issue seems to have disappeared. I will look into it a bit more. So leave this open a little longer and I will close it if I can't come up with more information.

The issue was only present in Firefox. Chrome worked fine.

**Comment by kevin-brown** [comment]
> So leave this open a little longer and I will close it if I can't come up with more information.

I think 2 months is good enough.

Closing this off since it cannot be reproduced.

**Comment by gastlich** [comment]
I can confirm the same issue is happening to me on Firefox mobile 68.0.2 described by @Bllacky

Didn't know that it was affected by wrongly positioned "touch" event. I thought it was always selecting previous option from the list.

**Comment by Bllacky** [comment]
It's definitely back. I can confirm it as well.

**Comment by gastlich** [comment]
@kevin-brown Would it be possible to reopen this issue, please? Additionally is there anything what we can do, to provide enough information for debugging.

I double checked on different Android device and behaves the same as described previously.

**Comment by kevin-brown** [comment]
I need a clear set of steps to reproduce this issue, since I tried it again on the same device and still can't reproduce it on the examples page.

**Comment by Bllacky** [comment]
Use an Android device, I have tried it with Android version 9.x and 10.x.
Use Firefox mobile from the PlayStore page.
Go to: https://select2.org/data-sources/ajax
Type "select2" in the Ajax Remote field.
Select second option - "react-component/select"

Select2 will select the option above it instead - "select2/select2"..

All my dev team members can reproduce this problem on their Android device using Firefox mobile.

**Comment by kevin-brown** [comment]
I was able to reproduce this yesterday and I may have narrowed down the specific case where this happens.

1. Go to the AJAX examples page
2. Open the dropdown
3. Search for anything that returns multiple results
4. Select an option other than the first result

The consistent behaviour is that the first result is selected, regardless of the position of the result you were trying for. During this testing, I also found that if you scroll the results before selecting them, this issue does not occur.

If someone else can confirm this narrowing of the case on their device, that would be useful.

**Comment by gastlich** [comment]
I can confirm, that the steps described by @kevin-brown are accurate. Additionally, scrolling fixes the "touch" selection issue on my devices also.

The only thing, which I'm not sure about is:
> The consistent behaviour is that the first result is selected, regardless of the position of the result you were trying for

In my case (as mentioned by @Bllacky ), it selects the previous option in the results, not always the very first one.

**Comment by kevin-brown** [comment]
Yeah, I actually can no longer reproduce that part and I'm back to it selecting the one before it. I can also confirm the weird part where it only appears to be an issue after the results are filtered.

On to the questions for narrowing this down

1. Has anyone tried to reproduce this using the "touch mode" of Firefox Desktop?
2. Is it possible that the "buffer space" between elements is correlated to the size of the search box in the dropdown? (this can be tested by making the search box large via CSS)
3. Does this issue happen in a multiple select dropdown? The documentation only has examples with a single select

**Comment by Bllacky** [comment]
1. I have tried multiple times but cannot reproduce it on FireFox desktop in "touch mode" or no "touch model".
3. Yes, it happens on multiple select dropdowns. If it helps, it happens on any of the drop-down selects here: https://noteb.com/?search/adv_search.php

2. Have no idea. Will try.

**Comment by gastlich** [comment]
> Does this issue happen in a multiple select dropdown? The documentation only has examples with a single select

Yes, I'm using only multiple select dropdowns (tagging support).

**Comment by kevin-brown** [comment]
Spent some more time on this.

It does not appear to be related to the search box. I tried to strip down the AJAX example in the documentation and it's not the following:

* Removing the placeholder still causes the issue
* Removing the `minimumInputLength` still causes the issue
* Setting the templating to only return the text doesn't fix it
  * Clicking on the avatar appears to consistently work
* It only happens when the AJAX data adapter or infinite scrolling is in use (Array data doesn't cause it)
* Removing the AJAX delay doesn't help
* Only displaying a single page of results (so `results.more = false`) doesn't fix the issue

It probably has to do with infinite scrolling.

**Comment by Bllacky** [comment]
This should not go stale again.

**Comment by kevin-brown** [comment]
> Maybe it's a FF bug?

My guess is that, but I haven't seen anything on the tracker about it and I haven't figured out the specific thing that is causing it.

**Comment by Bllacky** [comment]
> 
> 
> > Maybe it's a FF bug?
> 
> My guess is that, but I haven't seen anything on the tracker about it and I haven't figured out the specific thing that is causing it.

Maybe we should open a ticket on Bugzilla ? I will look again to see if there is anything related to this bug on their tracker.

**Comment by Bllacky** [comment]
Possibly related to:
https://bugzilla.mozilla.org/show_bug.cgi?id=1046017
https://bugzilla.mozilla.org/show_bug.cgi?id=1378431
https://bugzilla.mozilla.org/show_bug.cgi?id=1380447

Apparently this bug is only when in fullscreen mode, which I suppose it's all the time on an Android device.

We need to open a ticket with Mozilla.

**Comment by mike16889** [comment]
I have been having this issue in firefox on desktop. I found it was only showing itself on one select, not all of them. On said select, I had a placeholder item "Please make a selection" this one option had both attributes "selected" and "disabled". Removing "disabled" cleared the issue and allows it to work. Not optimal but will work for now.

**Comment by mbudde** [comment]
I've been trying to debug this issue on our site and found the following:

- When the callback for the `mouseup` event is called selecting a result, "Searching..." is visible as the first element in the search result list even though it was hidden before the click. This causes the rest of the results to be pushed down. The event target points to the result above the one clicked on. https://github.com/select2/select2/blob/e5131d0cc8dffd98ba36a68f3d027bf79e763cb4/src/js/select2/results.js#L444
- If the result list is scrolled down before selecting the result, the list result under the finger doesn't change when the "Searching..." element is displayed. In other words, the "Searching..." element is added above the scroll viewport. This means the `mouseup` event is triggered on the correct element.
- The reason the "Searching..." element is shown seems to be because an `input` event is triggered when selecting the result, just before the `mouseup` event triggers. https://github.com/select2/select2/blob/e5131d0cc8dffd98ba36a68f3d027bf79e763cb4/src/js/select2/selection/search.js#L136
- Setting `disableInputEvents` to true seems to make the bug go away.
https://github.com/select2/select2/blob/e5131d0cc8dffd98ba36a68f3d027bf79e763cb4/src/js/select2/selection/search.js#L113

**Comment by kevin-brown** [comment]
@mbudde that's definitely useful information to work from, thank you for doing some more digging into this issue!

Any chance you can provide info on the browser versions you tested with as well as the Select2 configuration? I'm still trying to figure out why we could only reproduce it on ajax configurations before.

**Comment by mbudde** [comment]
Here's a small example where I can reliably reproduce the bug:

https://jsbin.com/jadiwaqiye

Note that you need to enter something in the input field for the bug to trigger. In the plain select2 without the ajax configuration I cannot reproduce the bug. It seems `hideLoading` is called before the `mouseup` event handler while in the ajax version `hideLoading` is called after the `mouseup` event handler. I'm not sure why that is.

Tested with:
- Firefox Beta 68.6b4 on Android
- Firefox Preview Nightly 75.0a1 200302 on Android

I cannot reproduce in Firefox 74 on Linux.

**Comment by mbudde** [comment]
> It seems `hideLoading` is called before the `mouseup` event handler while in the ajax version `hideLoading` is called after the `mouseup` event handler. I'm not sure why that is.

Of course... the `input` event triggers an Ajax request which finishes after the `mouseup` event triggers.

**Comment by mbudde** [comment]
I think we're hitting this bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1610083

**Comment by kevin-brown** [comment]
That bug appears to be a duplicate of https://bugzilla.mozilla.org/show_bug.cgi?id=1305387 which should be reclassified as "clearly not going to be fixed" since it's been open for 4 years.

I'm tempted to suggest we should disable input event handling for Firefox for Android but that just feels wrong.

----

> I'm still trying to figure out why we could only reproduce it on ajax configurations before.

This is because infinite scroll is only enabled on AJAX configurations. Something which I realized almost immediately after posting that comment.

---

@mbudde I just wanted to say thank you for all of the debugging you have done. At this point we seem to have a clear idea of what the issue is. Now we just need to figure out how we want to work around it while we wait for Mozilla to fix the root cause of the issue.

**Comment by lazka** [comment]
Still a problem with Firefox 83 on Android.

Does anyone know of a workaround in the meantime?

**Comment by mahdiarjangi** [comment]
i had some issue with select2 search input, in iphone devices when i tapped select element the keyboard automatically popped-up and i add codes blow for solving the problem.

```javascript
const deviceInfo = {
  iPad: /iPad/.test(navigator.userAgent),
  iPhone: /iPhone/.test(navigator.userAgent),
};

$(".select_2").one("select2:open", function () {
  if (!Object.keys(deviceInfo).some((key) => deviceInfo[key])) {
    $(".select2-search input").attr("readonly", true);

    $(document).on("click", ".select2-search input", function (e) {
      e.preventDefault();
      $(this).attr("readonly", false);
    });
  }
});
```

**Comment by chrisjdixon** [comment]
@alexweissman if you're managing the docs could you please consider placing a note about Firefox mobile not working properly with this on [the Ajax docs page](https://select2.org/data-sources/ajax), despite this not looking like Select2's fault?

I've spent hours of frustration only to find this thread and my day is ruined. It'd be nice to help another person avoid this.

**Comment by SlyAceZeta** [comment]
I'm on Firefox for Android 124 and this looks to be resolved.

**Comment by Bllacky** [comment]
I will also test now after @SlyAceZeta comments, and I cannot reproduce it either.

**Comment by Bllacky** [comment]
If we have a third independent confirmation, I will close this issue.

**Comment by lazka** [comment]
I can also confirm that it's fixed.
